src/toil/jobStores/azureJobStore.py

```diff
@@ -800,12 +842,27 @@ def attempt(last=False):
         else:
             go.pop()
 
+    total_tries = num_tries
+    totalDelay = 0
+    delay = 0
     while go:
-        if num_tries == 1:
-            yield attempt(last=True)
+        try:
+            delay = waitTimes[total_tries - num_tries]
+        except IndexError:
+            delay = waitTimes[-1]
+        finally:
+            totalDelay += delay
+
+        if totalDelay >= timeout:
+            break
         else:
-            yield attempt()
+            time.sleep(delay)
+            yield attempt(last=num_tries == 1)
+
         # It's safe to do this, even with Python's weird default
         # arguments behavior, since we are assigning to num_tries
         # rather than mutating it.
         num_tries -= 1
+
+        if totalDelay >= timeout:
```

Comment:
If the attempts themselves take a long time your current timeout logic won't work. It will exceed the timeout considerably. Abort the loop if the **current** time plus the current delay exceeds the deadline. Compute the dead line at the beginning by adding the timeout to the current time. For reference look the retry logic in the AWS jobs store. You can basically copy it. The original author of this job store copied my the AWS implementation but then modified it with the intent to simplify it but the result doesn't handle the corner case as well.